src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java

```diff
@@ -0,0 +1,610 @@
+   * @return distance to the target in meters
+   */
+  public double getDistanceFromTargetMeters(double targetHeight) {
+    double angle = kCameraAngleDegrees + getOffSetY();
+    if (angle < 1 || angle > 89) {
+      return 0;
+    }
+    double tanTerm = Math.tan(Math.toRadians(angle));
+    double distance = (targetHeight - kCameraHeight) / tanTerm;
+
+    return distance;
+  }
+
+  /**
+   * Get the lateral distance to the target as defined in the construtor.
+   *
+   * @return distance to the target in meters
+   */
+  public double getDistanceFromTargetMeters() {
+    return getDistanceFromTargetMeters(kTargetHeight);
+  }
+
+  /**
+   * Get a 2d translation from the camera to the target, including normalization to handle the
+   * effects of angle to target. See the below discussion.
+   * https://www.chiefdelphi.com/t/what-does-limelight-skew-actually-measure/381167/7
+   *
+   * @parm targetHeightMeters use a target height other than what is provided in constructor.
+   * @return Translation2d from the camera to the target. This is camera relative, x is out of the
+   *     camera and positive y is to the left from the camera's point of view.
+   */
+  public Translation2d getTargetTranslation(double targetHeightMeters) {
+    /**
+     * This function uses the limelight coordinates until the end of the function. That is, the x
+     * axis is left/right, horizontal to the field. The y axis is 'up/down' normal to the field
+     * surface, and z is away from the camera, horizontal to the field surface.
+     *
+     * <p>The output coordinates are only a translation in x/y to the target, where x is out of the
+     * camera, horizontal to the field, and y is positive to the left from the perspective of the
+     * camera.
+     */
+    double diff_height = targetHeightMeters - kCameraHeight;
+
+    // x is the "left/right" to the target from the camera
+    double x = Math.tan(Math.toRadians(getOffSetX()));
+
+    // y is the 'up/down' to the target from the center of the camera
+    double y = Math.tan(Math.toRadians(getOffSetY()));
+
+    // z is straight out of the camera
+    double z = 1.0;
+    double length = Math.sqrt(x * x + y * y + z * z);
+
+    // Normalized vector components
+    double x_norm = x / length;
+    double y_norm = y / length;
+    double z_norm = z / length;
+
+    // Rotate vector by camera degrees around camera x axis
+    Translation2d zy_translation =
+        new Translation2d(z_norm, y_norm).rotateBy(Rotation2d.fromDegrees(kCameraAngleDegrees));
+    z_norm = zy_translation.getX();
+    y_norm = zy_translation.getY();
+
+    /**
+     * Find the intersection between the target in space, and the vector pointing to the target
+     *
+     * <p>This becomes a vector [x_norm, y_norm, z_norm] * d = [x_target, diff_height, z_target]
+     *
+     * <p>x_norm * d = x_target y_norm * d = diff_height z_target * d = z_target
+     */
+    double scaling = diff_height / y_norm;
```

Comment:
Generally you can just check if it's within some epsilon of zero and if yes, then replace it with a very small but non-zero value (like 1e-4 or something)

Reply:
Even if it's 1e-7 that's bad

Reply:
What do you mean? Even if `y_norm` is 1e-7? I don't see why that would be an issue, it should divide fine (it'll mean `scaling` is giant, but it would also be giant if I did 1e-4 like you suggested earlier).